src/MassTransit/DependencyInjection/DependencyInjection/ScopedMediator.cs

```diff
@@ -0,0 +1,248 @@
+        }
+
+        public RequestHandle<T> CreateRequest<T>(ConsumeContext consumeContext, Uri destinationAddress, object values,
+            CancellationToken cancellationToken = default,
+            RequestTimeout timeout = default)
+            where T : class
+        {
+            return _clientFactory.CreateRequest<T>(consumeContext, destinationAddress, values, cancellationToken, timeout);
+        }
+
+        public IRequestClient<T> CreateRequestClient<T>(RequestTimeout timeout = default)
+            where T : class
+        {
+            return _clientFactory.CreateRequestClient<T>(timeout);
+        }
+
+        public IRequestClient<T> CreateRequestClient<T>(ConsumeContext consumeContext, RequestTimeout timeout = default)
+            where T : class
+        {
+            return _clientFactory.CreateRequestClient<T>(consumeContext, timeout);
+        }
+
+        public IRequestClient<T> CreateRequestClient<T>(Uri destinationAddress, RequestTimeout timeout = default)
+            where T : class
+        {
+            return _clientFactory.CreateRequestClient<T>(destinationAddress, timeout);
+        }
+
+        public IRequestClient<T> CreateRequestClient<T>(ConsumeContext consumeContext, Uri destinationAddress, RequestTimeout timeout = default)
+            where T : class
+        {
+            return _clientFactory.CreateRequestClient<T>(consumeContext, destinationAddress, timeout);
+        }
+
+        public ConnectHandle ConnectConsumePipe<T>(IPipe<ConsumeContext<T>> pipe)
+            where T : class
+        {
+            return _mediator.ConnectConsumePipe(pipe);
+        }
+
+        public ConnectHandle ConnectConsumePipe<T>(IPipe<ConsumeContext<T>> pipe, ConnectPipeOptions options)
+            where T : class
+        {
+            return _mediator.ConnectConsumePipe(pipe, options);
+        }
+
+        public ConnectHandle ConnectRequestPipe<T>(Guid requestId, IPipe<ConsumeContext<T>> pipe)
+            where T : class
+        {
+            return _mediator.ConnectRequestPipe(requestId, pipe);
+        }
+
+        public ConnectHandle ConnectConsumeObserver(IConsumeObserver observer)
+        {
+            return _mediator.ConnectConsumeObserver(observer);
+        }
+
+        public ConnectHandle ConnectConsumeMessageObserver<T>(IConsumeMessageObserver<T> observer)
+            where T : class
+        {
+            return _mediator.ConnectConsumeMessageObserver(observer);
+        }
+
+        protected override IPipe<SendContext<T>> GetPipeProxy<T>(IPipe<SendContext<T>> pipe = default)
+        {
+            return new ScopedSendPipeAdapter<T>(_provider, pipe);
+        }
+
+        async Task PublishInternal<T>(CancellationToken cancellationToken, T message, IPipe<PublishContext<T>> pipe = default)
+            where T : class
+        {
+            var sendEndpoint = await GetPublishSendEndpoint<T>().ConfigureAwait(false);
```

Comment:
Copy [the same behavior](https://github.com/MassTransit/MassTransit/blob/8a72e940a70433453b7edbc554b662ce70c49937/src/MassTransit.Abstractions/Transports/PublishEndpoint.cs#L112) to avoid all the async state machines.